tests/test_random_number_generator.h

```diff
@@ -106,6 +105,110 @@ TEST_CASE("[RandomNumberGenerator] Restore from seed") {
+	float num_before;
+	for (int i = 0; i < 1000; i++) {
+		num_before = rng->randf_range(-100, 100);
+		CHECK(num_before >= -100);
+		CHECK(num_before <= 100);
+	}
+
+	float num_after1;
+	rng->randomize();
+	INFO("randf_range should give float between -75 and 75.");
+	INFO("Shouldn't be affected by randomize.");
+	for (int i = 0; i < 1000; i++) {
+		num_after1 = rng->randf_range(-75, 75);
+		CHECK(num_after1 >= -75);
+		CHECK(num_after1 <= 75);
+	}
+
+	rng->set_state(prev_state);
+	int num_after2;
+	INFO("randi_range should give int between -50 and 50.");
+	INFO("Shouldn't be affected by set_state.");
+	for (int i = 0; i < 1000; i++) {
+		num_after2 = rng->randi_range(-50, 50);
+		CHECK(num_after2 >= -50);
+		CHECK(num_after2 <= 50);
+	}
+
+	rng->set_seed(seed_before);
+	int num_after3;
+	INFO("randi_range should give int between -25 and 25.");
+	INFO("Shouldn't be affected by set_seed.");
+	for (int i = 0; i < 1000; i++) {
+		num_after3 = rng->randi_range(-25, 25);
+		CHECK(num_after3 >= -25);
+		CHECK(num_after3 <= 25);
+	}
+
+	rng->randf();
+	rng->randf();
+	float num_after4;
+
+	INFO("randf_range should give float between -10 and 10.");
+	INFO("Shouldn't be affected after generating new numbers.");
+	for (int i = 0; i < 1000; i++) {
+		num_after4 = rng->randf_range(-10, 10);
+		CHECK(num_after4 >= -10);
+		CHECK(num_after4 <= 10);
+	}
+
+	rng->randi();
+	rng->randi();
+	int num_after5;
+
+	INFO("randi_range should give int between -5 and 5.");
+	INFO("Shouldn't be affected after generating new numbers.");
+	for (int i = 0; i < 1000; i++) {
+		num_after5 = rng->randf_range(-5, 5);
+		CHECK(num_after5 >= -5);
+		CHECK(num_after5 <= 5);
+	}
+}
+
+TEST_CASE("[RandomNumberGenerator] Check randfn for expected functionality") {
+	Ref<RandomNumberGenerator> rng = memnew(RandomNumberGenerator);
+	rng->set_seed(1);
+	INFO("randfn should give a number between -4 to 4 (4 std deviations away; above 99.7% chance it will be in this range).");
+	INFO("Standard randfn function call.");
+	for (int i = 0; i < 1000; i++) {
+		CHECK(rng->randfn() >= -4);
+		CHECK(rng->randfn() <= 4);
+	}
+
+	INFO("Checks if randfn produces number between -3 to 3 after multiple randi/randf calls.");
+	INFO("4 std deviations away; above 99.7% chance it will be in this range.");
+	rng->randf();
+	rng->randi();
+	for (int i = 0; i < 1000; i++) {
+		CHECK(rng->randfn() >= -4);
+		CHECK(rng->randfn() <= 4);
+	}
+
+	INFO("Checks if user defined mean and deviation work properly.");
+	INFO("4 std deviations away; above 99.7% chance it will be in this range.");
+	for (int i = 0; i < 1000; i++) {
+		CHECK(rng->randfn(5, 10) >= -35);
+		CHECK(rng->randfn(5, 10) <= 45);
+	}
+
+	INFO("Checks if randfn works with changed seeds.");
+	INFO("4 std deviations away; above 99.7% chance it will be in this range.");
+	rng->randomize();
+	for (int i = 0; i < 1000; i++) {
+		CHECK(rng->randfn(3, 3) >= -9);
+		CHECK(rng->randfn(3, 3) <= 15);
+	}
```

Comment:
Are those meant not to be centered on 0?

And what does "above 99.7% chance it will be in this range" mean for the test, that it will fail 0.3% of the times? With 1000 iterations that means a high chance of failure, and tests should never fail if not broken.

Reply:
In terms of Gaussian distribution according to the Godot Random Number Generator documentation, 99.7% of the numbers in the distribution would be covered 3 standard deviations from the mean. I believe I used 4 standard deviations to get as close to 100%. From my understanding of the documentation, the center point would be the first argument of the randfn() function as the distribution would be centered around mean. 

In hindsight, I see that there was no point in testing 1,000 times for something like this since there's still a chance of failure. Do you have any recommendations on how many iterations should be done in order for the test to be considered rigorous?

Reply:
Given the same seed, you could decrease the number of iterations to like 100 perhaps, so that it would indeed succeed in those 99.7% cases, or try to pick a different seed instead, unfortunately I'm not experienced with Gaussian distribution to elaborate further.

The downside is that if we change the RNG implementation at some point, the test could fail and then have to be updated again, but that's the point of unit testing!

Reply:
Yea no worries, I fixed the number of iterations and also increased the number of std deviations to 5 so that test failure is less likely to occur now.

Reply:
Well if it can be outside the range, the test should not expect it to be in the range, otherwise it's always bound to fail at some point. Reducing the number of iterations to reducing the likelihood of failure will just mean that failures will happen ten time less, or depends on internal RNG features that could guarantee that the fake randomness actually passes the test... but in any case it's brittle and not very useful.

A more useful test would be to check if we do have over 99.7% of values in the expected range.